crates/stress-test/src/main.rs

```diff
@@ -0,0 +1,176 @@
+
+    #[arg(short('O'), long)]
+    /// Show the container output in stdout
+    container_output: bool,
+
+    /// Path to the shim binary
+    shim: PathBuf,
+
+    #[arg(short, long, default_value("1"))]
+    /// Number of tasks to create and start concurrently [0 = no limit]
+    parallel: usize,
+
+    #[arg(short('S'), long, default_value("start"))]
+    /// Up to what steps to run in series
+    serial_steps: Step,
+
+    #[arg(short('n'), long, default_value("10"))]
+    /// Number of tasks to run
+    count: usize,
+
+    #[clap(short, long, value_parser = humantime::parse_duration, default_value = "2s")]
+    /// Runtime timeout [0 = no timeout]
+    timeout: Duration,
+}
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<()> {
+    set_child_subreaper(true)?;
+    let res = main_impl().await;
+    let _ = reap_children().await?;
+    res
+}
+
+async fn main_impl() -> Result<()> {
+    env_logger::try_init()?;
+
+    let Cli {
+        shim,
+        verbose,
+        container_output,
+        parallel,
+        serial_steps,
+        count,
+        timeout,
+    } = Cli::parse();
+    let shim: String = shim.to_string_lossy().into_owned();
+
+    let c8d = mocks::Containerd::new(verbose).await?;
+
+    let shim = c8d.start_shim(shim).await?;
+    let shim = Arc::new(shim);
+
+    let permits = if parallel == 0 {
+        count as _
+    } else {
+        parallel as _
+    };
+    let semaphore = Arc::new(Semaphore::new(permits));
+    let mut tracker: JoinSet<Result<()>> = JoinSet::new();
+
+    let serial_steps = match serial_steps {
+        Step::Create => "c",
+        Step::Start => "cs",
+        Step::Wait => "csw",
+        Step::Delete => "cswd",
+    };
+
+    for _ in 0..count {
+        let shim = shim.clone();
+        let semaphore = semaphore.clone();
+
+        tracker.spawn(async move {
+            let task = shim.task().await?;
+
+            {
+                // take a permit as this saction might have to run serially
+                let _permit = semaphore.acquire().await?;
+                if serial_steps.contains('c') {
+                    task.create(container_output).await?;
+                }
+                if serial_steps.contains('s') {
+                    task.start().await?;
+                }
+                if serial_steps.contains('w') {
+                    task.wait().await?;
+                }
+                if serial_steps.contains('d') {
+                    task.delete().await?;
+                }
+            }
+
+            if !serial_steps.contains('c') {
+                task.create(container_output).await?;
+            }
+            if !serial_steps.contains('s') {
+                task.start().await?;
+            }
+            if !serial_steps.contains('w') {
+                task.wait().await?;
+            }
+            if !serial_steps.contains('d') {
+                task.delete().await?;
+            }
+
+            Ok(())
+        });
+    }
```

Comment:
```suggestion
for _ in 0..count {
        let serial_steps = match serial_steps {
            Step::Create => vec![],
            Step::Start => vec![Step::Start],
            Step::Wait => vec![Step::Start, Step::Wait],
            Step::Delete => vec![Step::Start, Step::Wait, Step::Delete],
        };
        let shim = shim.clone();
        let semaphore = semaphore.clone();

        tracker.spawn(async move {
            let task = shim.task().await?;

            {
                // take a permit as this saction might have to run serially
                let _permit = semaphore.acquire().await?;
                task.create(container_output).await?;
                if serial_steps.contains(&Step::Start) {
                    task.start().await?;
                }
                if serial_steps.contains(&Step::Wait) {
                    task.wait().await?;
                }
                if serial_steps.contains(&Step::Delete) {
                    task.delete().await?;
                }
            }
            if !serial_steps.contains(&Step::Start) {
                task.start().await?;
            }
            if !serial_steps.contains(&Step::Wait) {
                task.wait().await?;
            }
            if !serial_steps.contains(&Step::Delete) {
                task.delete().await?;
            }

            Ok(())
        });
    }
```

Reply:
hm, I think I can derive PartialOrd on the enum, and use that, I'll give that a go.

Reply:
Ok, I came up with a different solution, let me know what you think.

Reply:
Feels much better now